File: graph_components/dfs/src/DfsDemo.swift
// Swift
// グラフの連結成分: DFS

class GraphData {
    // 隣接ノードとその辺の重みを格納します。
    // キーは頂点、値はその頂点に隣接する頂点と重みの配列です。
    private var _data: [String: [(String, Int)]] = [:]
    
    init() {}
    
    func get() -> [String: [(String, Int)]] {
        // グラフの内部データを取得します。
        return _data
    }
    
    func getVertices() -> [String] {
        // グラフの全頂点をリストとして返します。
        return Array(_data.keys)
    }
        
    func addVertex(vertex: String) -> Bool {
        // 新しい頂点をグラフに追加します。
        if _data[vertex] == nil {
            _data[vertex] = []
        }
        return true
    }
    
    func addEdge(vertex1: String, vertex2: String, weight: Int) -> Bool {
        // 両頂点間に辺を追加します。重みを指定します。
        // 頂点がグラフに存在しない場合は追加します。
        if _data[vertex1] == nil {
            _ = addVertex(vertex: vertex1)
        }
        if _data[vertex2] == nil {
            _ = addVertex(vertex: vertex2)
        }
        
        // vertex1 -> vertex2 の辺を追加（重み付き）
        var edgeExistsV1V2 = false
        if var neighbors = _data[vertex1] {
            for (index, (neighbor, _)) in neighbors.enumerated() {
                if neighbor == vertex2 {
                    neighbors[index] = (vertex2, weight) // 既に存在する場合は重みを更新
                    _data[vertex1] = neighbors
                    edgeExistsV1V2 = true
                    break
                }
            }
            if !edgeExistsV1V2 {
                _data[vertex1]?.append((vertex2, weight))
            }
        }
        
        // vertex2 -> vertex1 の辺を追加（重み付き）
        var edgeExistsV2V1 = false
        if var neighbors = _data[vertex2] {
            for (index, (neighbor, _)) in neighbors.enumerated() {
                if neighbor == vertex1 {
                    neighbors[index] = (vertex1, weight) // 既に存在する場合は重みを更新
                    _data[vertex2] = neighbors
                    edgeExistsV2V1 = true
                    break
                }
            }
            if !edgeExistsV2V1 {
                _data[vertex2]?.append((vertex1, weight))
            }
        }
        
        return true
    }
       
    func clear() -> Bool {
        // グラフを空にする
        _data = [:]
        return true
    }
    
    private func _dfs(vertex: String, visited: inout Set<String>, currentComponent: inout [String]) {
        // 現在の頂点を訪問済みにマークし、現在の成分に追加
        visited.insert(vertex)
        currentComponent.append(vertex)
        
        // 隣接する頂点を探索
        guard let neighbors = _data[vertex] else { return }
        
        for neighborInfo in neighbors {
            let neighborVertex = neighborInfo.0
            
            // まだ訪問していない隣接頂点に対して再帰的にDFSを呼び出す
            if !visited.contains(neighborVertex) {
                _dfs(vertex: neighborVertex, visited: &visited, currentComponent: &currentComponent)
            }
        }
    }
    
    func getConnectedComponents() -> [[String]] {
        var visited = Set<String>() // 訪問済み頂点を記録するセット
        var connectedComponents: [[String]] = [] // 連結成分を格納するリスト
        
        // グラフの全頂点を順にチェック
        for vertex in getVertices() {
            // まだ訪問していない頂点からDFSを開始
            if !visited.contains(vertex) {
                var currentComponent: [String] = [] // 現在の連結成分を格納するリスト
                // DFSヘルパー関数を呼び出し、現在の連結成分を探索
                _dfs(vertex: vertex, visited: &visited, currentComponent: &currentComponent)
                // 探索で見つかった連結成分を結果リストに追加
                connectedComponents.append(currentComponent)
            }
        }
        
        return connectedComponents
    }
}

func main() {
    print("Dfs TEST -----> start")
    
    print("\nnew")
    let graphData = GraphData()
    print("  現在のデータ: \(graphData.get())")
    
    print("\nadd_edge")
    _ = graphData.clear()
    let inputList1: [(String, String, Int)] = [("A", "B", 4), ("B", "C", 3), ("B", "D", 2), ("D", "A", 1), ("A", "C", 2), ("B", "D", 2)]
    for input in inputList1 {
        print("  入力値: \(input)")
        let output = graphData.addEdge(vertex1: input.0, vertex2: input.1, weight: input.2)
        print("  出力値: \(output)")
    }
    print("  現在のデータ: \(graphData.get())")
    print("\nget_connected_components")
    let output1 = graphData.getConnectedComponents()
    print("  連結成分: \(output1)")
    
    print("\nadd_edge")
    _ = graphData.clear()
    let inputList2: [(String, String, Int)] = [("A", "B", 4), ("C", "D", 4), ("E", "F", 1), ("F", "G", 1)]
    for input in inputList2 {
        print("  入力値: \(input)")
        let output = graphData.addEdge(vertex1: input.0, vertex2: input.1, weight: input.2)
        print("  出力値: \(output)")
    }
    print("  現在のデータ: \(graphData.get())")
    print("\nget_connected_components")
    let output2 = graphData.getConnectedComponents()
    print("  連結成分: \(output2)")
    
    print("\nadd_edge")
    _ = graphData.clear()
    let inputList3: [(String, String, Int)] = [("A", "B", 4), ("B", "C", 3), ("D", "E", 5)]
    for input in inputList3 {
        print("  入力値: \(input)")
        let output = graphData.addEdge(vertex1: input.0, vertex2: input.1, weight: input.2)
        print("  出力値: \(output)")
    }
    print("  現在のデータ: \(graphData.get())")
    print("\nget_connected_components")
    let output3 = graphData.getConnectedComponents()
    print("  連結成分: \(output3)")
    
    print("\nadd_edge")
    _ = graphData.clear()
    let inputList4: [(String, String, Int)] = []
    for input in inputList4 {
        print("  入力値: \(input)")
        let output = graphData.addEdge(vertex1: input.0, vertex2: input.1, weight: input.2)
        print("  出力値: \(output)")
    }
    print("  現在のデータ: \(graphData.get())")
    print("\nget_connected_components")
    let output4 = graphData.getConnectedComponents()
    print("  連結成分: \(output4)")
    
    print("Dfs TEST <----- end")
}

main()

File: graph_components/dfs/src/DfsDemo.rb
# Ruby
# グラフの連結成分: DFS

class GraphData
  def initialize
    # 隣接ノードとその辺の重みを格納します。
    # キーは頂点、値はその頂点に隣接する頂点と重みの配列です。
    # 例: { 'A' => [['B', 1], ['C', 4]], 'B' => [['A', 1], ['C', 2], ['D', 5]], ... }
    @data = {}
  end

  def get
    # グラフの内部データを取得します。
    @data
  end

  def get_vertices
    # グラフの全頂点をリストとして返します。
    @data.keys
  end
  
  def add_vertex(vertex)
    # 新しい頂点をグラフに追加します。
    unless @data.key?(vertex)
      @data[vertex] = []
      return true
    end
    # 既に存在する場合は追加しないがtrueを返す（変更なしでも成功とみなす）
    true
  end
  
  def add_edge(vertex1, vertex2, weight)
    # 両頂点間に辺を追加します。重みを指定します。
    # 頂点がグラフに存在しない場合は追加します。
    add_vertex(vertex1) unless @data.key?(vertex1)
    add_vertex(vertex2) unless @data.key?(vertex2)
    
    # 両方向に辺を追加する（無向グラフ）
    # 既に同じ辺が存在する場合は重みを更新するかどうか？
    # 今回は単純に追加する（重複辺はプリム法では問題にならないが、データとしては綺麗でないかも）
    # ここでは、同じ頂点間の辺が存在しない場合のみ追加するように修正します。
    
    # vertex1 -> vertex2 の辺を追加（重み付き）
    edge_exists_v1v2 = false
    @data[vertex1].each_with_index do |(neighbor, _), i|
      if neighbor == vertex2
        @data[vertex1][i] = [vertex2, weight] # 既に存在する場合は重みを更新
        edge_exists_v1v2 = true
        break
      end
    end
    @data[vertex1] << [vertex2, weight] unless edge_exists_v1v2

    # vertex2 -> vertex1 の辺を追加（重み付き）
    edge_exists_v2v1 = false
    @data[vertex2].each_with_index do |(neighbor, _), i|
      if neighbor == vertex1
        @data[vertex2][i] = [vertex1, weight] # 既に存在する場合は重みを更新
        edge_exists_v2v1 = true
        break
      end
    end
    @data[vertex2] << [vertex1, weight] unless edge_exists_v2v1
    
    true
  end
    
  def clear
    # グラフを空にする
    @data = {}
    true
  end

  def _dfs(vertex, visited, current_component)
    # 現在の頂点を訪問済みにマークし、現在の成分に追加
    visited.add(vertex)
    current_component << vertex

    # 隣接する頂点を探索
    # @data[vertex] || [] は、vertexが存在しない場合でも
    # エラーにならず空配列を返す安全な方法
    # ここで取得されるのは [[neighbor, weight], ...] の配列です。
    (@data[vertex] || []).each do |neighbor_info|
      # neighbor_info は [neighbor, weight] の配列なので、
      # 配列の0番目の要素（隣接頂点そのもの）を取得します。
      neighbor_vertex = neighbor_info[0]

      # まだ訪問していない隣接頂点に対して再帰的にDFSを呼び出す
      unless visited.include?(neighbor_vertex)
        _dfs(neighbor_vertex, visited, current_component)
      end
    end
  end

  def get_connected_components
    visited = Set.new # 訪問済み頂点を記録するセット
    connected_components = [] # 連結成分を格納する配列

    # グラフの全頂点を順にチェック
    get_vertices.each do |vertex|
      # まだ訪問していない頂点からDFSを開始
      unless visited.include?(vertex)
        current_component = [] # 現在の連結成分を格納する配列
        # DFSヘルパー関数を呼び出し、現在の連結成分を探索
        _dfs(vertex, visited, current_component)
        # 探索で見つかった連結成分を結果配列に追加
        connected_components << current_component
      end
    end

    connected_components
  end
end

def main
  puts "Dfs TEST -----> start"

  puts "\nnew"
  graph_data = GraphData.new
  puts "  現在のデータ: #{graph_data.get}"

  puts "\nadd_edge"
  graph_data.clear
  input_list = [['A', 'B', 4], ['B', 'C', 3], ['B', 'D', 2], ['D', 'A', 1], ['A', 'C', 2], ['B', 'D', 2]]
  input_list.each do |input|
    puts "  入力値: #{input}"
    output = graph_data.add_edge(*input)
    puts "  出力値: #{output}"
  end
  puts "  現在のデータ: #{graph_data.get}"
  puts "\nget_connected_components"
  output = graph_data.get_connected_components
  puts "  連結成分: #{output}"

  puts "\nadd_edge"
  graph_data.clear
  input_list = [['A', 'B', 4], ['C', 'D', 4], ['E', 'F', 1], ['F', 'G', 1]]
  input_list.each do |input|
    puts "  入力値: #{input}"
    output = graph_data.add_edge(*input)
    puts "  出力値: #{output}"
  end
  puts "  現在のデータ: #{graph_data.get}"
  puts "\nget_connected_components"
  output = graph_data.get_connected_components
  puts "  連結成分: #{output}"

  puts "\nadd_edge"
  graph_data.clear        
  input_list = [['A', 'B', 4], ['B', 'C', 3], ['D', 'E', 5]]
  input_list.each do |input|
    puts "  入力値: #{input}"
    output = graph_data.add_edge(*input)
    puts "  出力値: #{output}"
  end
  puts "  現在のデータ: #{graph_data.get}"
  puts "\nget_connected_components"
  output = graph_data.get_connected_components
  puts "  連結成分: #{output}"

  puts "\nadd_edge"
  graph_data.clear
  input_list = []
  input_list.each do |input|
    puts "  入力値: #{input}"
    output = graph_data.add_edge(*input)
    puts "  出力値: #{output}"
  end
  puts "  現在のデータ: #{graph_data.get}"
  puts "\nget_connected_components"
  output = graph_data.get_connected_components
  puts "  連結成分: #{output}"

  puts "Dfs TEST <----- end"
end

# Rubyの慣習に従い、Setクラスを使用するためにrequireを追加
require 'set'

# Rubyのメインプログラム実行
if __FILE__ == $PROGRAM_NAME
  main
end